### PR TITLE
feat(FXC-4519): Added check to warn of duplicated voltage values

### PR DESCRIPTION
**Pull request by marc-flex**
<h3>Greptile Summary</h3>

Added validation to `DCVoltageSource` that warns users when duplicate voltage values are detected in the `voltage` array. The implementation treats `0` and `-0` as the same value by normalizing zeros before checking for duplicates.

- Implemented `check_repeated_voltage` validator that normalizes zero values and uses `np.unique()` to detect duplicates
- Added comprehensive test coverage including edge cases for zero normalization and multiple duplicates
- Updated CHANGELOG.md with clear user-facing description of the new feature

**Note:** Previous thread comments identified a floating-point precision issue - the current implementation only normalizes zeros but uses direct equality for non-zero values via `np.unique()`, which may miss near-duplicate floating-point values like `1.0` and `1.0000000001`.

<h3>Confidence Score: 4/5</h3>

- This PR is safe to merge with minor style improvements recommended
- The implementation is functionally correct for the intended use case of detecting exact duplicates and normalizing zeros. Previous thread already identified the floating-point precision consideration for non-zero values. Test coverage is comprehensive. Only minor style improvement suggested for message formatting.
- Pay attention to `tidy3d/components/spice/sources/dc.py` regarding the floating-point comparison approach discussed in previous comments

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| tidy3d/components/spice/sources/dc.py | Added duplicate voltage validation with zero normalization. Previous thread identified floating-point precision issue that needs addressing. |
| tests/test_components/test_heat_charge.py | Comprehensive test coverage for duplicate voltage warning, including edge cases like 0/-0 and multiple duplicates. |
| CHANGELOG.md | Clear changelog entry in Added section describing the new validation feature. |

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant User
    participant DCVoltageSource
    participant Validator
    participant numpy as NumPy
    participant Logger

    User->>DCVoltageSource: Create with voltage array
    DCVoltageSource->>Validator: @validator("voltage") check_voltage
    Validator->>Validator: Check for infinite values
    Validator-->>DCVoltageSource: Return validated values
    
    DCVoltageSource->>Validator: @validator("voltage") check_repeated_voltage
    Validator->>numpy: np.isclose(val, 0, atol=1e-10)
    numpy-->>Validator: Boolean mask for zeros
    Validator->>numpy: np.where(mask, 0.0, val)
    numpy-->>Validator: Normalized array (zeros unified)
    Validator->>numpy: np.unique(normalized)
    numpy-->>Validator: Unique values array
    Validator->>Validator: Compare len(unique_values) < len(val)
    alt Duplicates found
        Validator->>Logger: log.warning(duplicate message)
        Logger-->>User: Display warning
    end
    Validator-->>DCVoltageSource: Return original values
    DCVoltageSource-->>User: Instance created
```

---

> [!NOTE]
> Introduces a tolerance-based warning for duplicate entries in `DCVoltageSource.voltage`, normalizing `0`/`-0` and logging a warning when repeats are detected.
> 
> - Adds `_count_unique_with_tolerance` and a `@validator` (`check_repeated_voltage`) in `dc.py` to detect near-duplicates via `np.isclose`, treating `0` and `-0` as the same
> - Extends tests with `test_repeated_voltage_warning` to assert warning behavior for unique values, exact duplicates, `0` vs `-0`, and multiple repeats
> - Updates `CHANGELOG.md` to document the new validation
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 80e8e39dcbb6f35fb215878c7cb0fdd360772f15. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Comment by marc-flex** [comment]
@greptile update summary

**Comment by marc-flex** [comment]
@greptile update summary

File: tidy3d/components/spice/sources/dc.py [hunk 1]
```diff
@@ -73,6 +75,20 @@ def check_voltage(cls, val):
...
+        """Warn if repeated voltage values are present, treating 0 and -0 as the same value."""
+        # Normalize all zero values (both 0.0 and -0.0) to 0.0 so they are treated as duplicates
+        normalized = np.where(np.isclose(val, 0, atol=1e-10), 0.0, val)
+        unique_values = np.unique(normalized)
+        if len(unique_values) < len(val):
+            log.warning(
+                "Duplicate voltage values detected in 'voltage' array. "
+                f"Found {len(val)} values but only {len(unique_values)} are unique. "
+                "Note: '0' and '-0' are considered the same value."
+            )
```
Comment (momchil-flex) [inline tidy3d/components/spice/sources/dc.py:110]: Generally looks good, but the one thing that bugs me a little is there's no guaranteed correspondence between what's validated here and what values the solver will compute at. Would be nice to have a single source of truth for the "computed voltages", which also e.g. defines the tolerance within which two values are considered identical.

**Review by momchil-flex — APPROVED** [review]
